File: 김태선/Week_11/P72411_메뉴 리뉴얼.py
from collections import Counter
from itertools import combinations


def solution(orders, course) : 
    answer = []

    for c_length in course : 
        temp = []

        for order in orders : 
            # 조합 라이브러리 함수(문자열 조합을 갯수에 맞게 계산)
            combi = combinations(sorted(order), c_length)
            # 문자열 조합을 temp에 저장
            temp += list(combi)

        # 계산 라이브러리 함수(딕셔너리 형태로 분류해서 총 가짓수를 구해줌)
        counter = Counter(temp)

        # 길이가 0 이상이고 가장 큰 조합 갯수가 최소 2번 이상인 값만 구해줌
        if len(counter) != 0 and max(counter.values()) != 1 :
            for c_factor in counter : 
                if counter[c_factor] == max(counter.values()) :
                    answer.append(''.join(c_factor)) # join 써서 문자열로 변환

    return sorted(answer)

File: 길민지/Week_11/P72411_메뉴리뉴얼.java
import java.util.*;
class Solution {
    
    public static Map<String, Integer>[] mapArr;
    public static int N;
    public static List<String> list;

    public static void sol(String[] orders, int[] course){

        for(int i=0; i<course.length; i++){
            int r = course[i];
            mapArr[i] = new HashMap<>();
            for(String order : orders){
            //nCr 구하기
            char[] c = new char[r];
            combination(order, 0,0, r, c, mapArr[i]);
            }
        }

        for(int i=0; i<course.length; i++){
            int max = 2;
            for(String key : mapArr[i].keySet()){
                max = Math.max(max, mapArr[i].get(key));
            }
            for(String key : mapArr[i].keySet()){
                if(mapArr[i].get(key)==max) list.add(key);
            }
        }
        Collections.sort(list);
    }
    
    public static void combination(String order, int i, int n, int r, char[] c, Map<String, Integer> map){ 
        // 코스, 배열 인덱스, 위치, 배열 크기, 배열
        if(i==r){
            char[] newArr = c.clone();
            Arrays.sort(newArr);
            String s = new String(newArr);
            if(map.containsKey(s)){
                map.put(s, map.get(s)+1);
            }
            else{
                map.put(s, 1);
            }
            return;
        }
        if(n>=order.length()) return;
        // 선택x
        combination(order, i, n+1, r, c, map);
        // 선택o
        c[i] = order.charAt(n);
        combination(order, i+1, n+1, r, c, map);
    }
    
    public String[] solution(String[] orders, int[] course) {
        mapArr = new Map[course.length];
        list = new ArrayList<>();
        sol(orders, course);
        
        System.out.println(list);
        
        String[] answer = list.stream()
                            .toArray(String[]::new);
        return answer;
    }
}



